src/main/starlark/builtins_bzl/common/cc/compile/compile.bzl

```diff
@@ -436,6 +493,602 @@ def _should_provide_header_modules(
+
+    # ctx = cc_internal.actions2ctx_cheat(actions)
+    args = actions.args()
+    args.add_all(ddi_files, before_each = "-d")
+    args.add_all(direct_module_files, before_each = "-f")
+    args.add_all(transitive_modules_info_files, before_each = "-m")
+    args.add("-o", modules_info_file)
+    actions.run(
+        outputs = [modules_info_file],
+        # inputs = depset(ddi_files + direct_module_files, transitive = [transitive_modules_info_files]),
+        inputs = depset(ddi_files, transitive = [transitive_modules_info_files]),
+        executable = aggregate_ddi_tool,
+        tools = [aggregate_ddi_tool],
+        arguments = [args],
+        mnemonic = "CppAggregateDdi",
+        progress_message = "Generating C++20 modules info %{output}",
+        toolchain = "@bazel_tools//tools/cpp:toolchain_type",
+    )
+
+def _create_gen_modmap_action(
+        *,
+        actions,
+        cc_toolchain,
+        ddi_file,
+        modules_info_file,
+        modmap_file,
+        modmap_input_file):
+    gen_modmap_tool = cc_toolchain.generate_modmap
+    if gen_modmap_tool == None:
+        fail("the 'generate_modmap' tool is not defined in the C++ toolchain")
+    actions.run(
+        outputs = [modmap_file, modmap_input_file],
+        inputs = [ddi_file, modules_info_file],
+        executable = gen_modmap_tool,
+        tools = [gen_modmap_tool],
+        arguments = [cc_toolchain.compiler, ddi_file.path, modules_info_file.path, modmap_file.path],
+        mnemonic = "CppGenModmap",
+        progress_message = "Generating C++20 modules modmap %{output}",
+        toolchain = "@bazel_tools//tools/cpp:toolchain_type",
+    )
+
+def _create_cc_compile_actions_with_cpp20_module_helper(
+        *,
+        actions,
+        action_construction_context,
+        additional_compilation_inputs,
+        additional_include_scanning_roots,
+        cc_compilation_context,
+        cc_toolchain,
+        compilation_unit_sources,
+        module_interfaces_sources,
+        configuration,
+        conlyopts,
+        copts,
+        copts_filter,
+        cpp_configuration,
+        cxxopts,
+        fdo_context,
+        feature_configuration,
+        is_code_coverage_enabled,
+        label,
+        private_headers,
+        public_headers,
+        purpose,
+        language,
+        outputs,
+        common_compile_build_variables,
+        auxiliary_fdo_inputs,
+        fdo_build_variables,
+        use_pic,
+        output_name_map):
+    direct_module_files = []
+    source_to_module_file_map = {}
+    source_to_ddi_file_map = {}
+    modules_info_file = _get_compile_output_file(
+        action_construction_context,
+        label,
+        output_name = cc_internal.get_artifact_name_for_category(
+            cc_toolchain = cc_toolchain,
+            category = artifact_category.CPP_MODULES_INFO,
+            output_name = label.name,
+        ),
+    )
+    if use_pic:
+        outputs.add_pic_modules_info_file(modules_info_file)
```

Comment:
You could add a `pic` `boolean` parameter to these functions, which cuts their number in half and also makes the callsite more convenient. I know the existing functions don't do that, but we don't have to add to that mess :-) 

Force the parameter to be named, not positional, so that the call sites are easy to read.